content/300-guides/300-upgrade-guides/200-upgrading-versions/030-upgrading-to-prisma-4.mdx

```diff
@@ -0,0 +1,418 @@
+  // HINT: No function matches the given name and argument types. You might need to add explicit type casts.
+```
+
+This is because PostgreSQL’s `length` function expects `text` as input. Prisma used to silently coerce `42` to `text`, but does not do this in version 4.0.0. To fix this, explicitly cast `42`to `text` as follows:
+
+```js
+await prisma.$queryRaw`select length(${42}::text);
+```
+
+### <inlineCode>DbNull</inlineCode>, <inlineCode>JsonNull</inlineCode> and <inlineCode>AnyNull</inlineCode> are now objects
+
+JavaScript `null` is ambiguous for JSON columns, so Prisma uses `DbNull`, `JsonNull`, and `AnyNull` to distinguish between the database `NULL` value and the JSON `null` value. Before version 4.0.0, `DbNull`, `JsonNull`, and `AnyNull` were string constants. From version 4.0.0, they are objects.
+
+See [Filtering by null values](/concepts/components/prisma-client/working-with-fields/working-with-json-fields#filtering-by-null-values) for more information.
+
+#### Upgrade path
+
+1. If you use literal strings to address these values, then you must replace them with the following named constants:
+
+   - `DbNull`: replace with `Prisma.DbNull`
+   - `JsonNull`: replace with `Prisma.JsonNull`
+   - `AnyNull`: replace with `Prisma.AnyNull`
+
+   If you already use these named constants, then you do not need to take any action.
+
+1. If you now get a type error when you pass `Prisma.DbNull` as the value of a JSON field, then this probably indicates a bug in your code that our types did not catch before version 4.0.0. The field where you tried to store `DbNull` is probably not nullable in your schema. As a result, a literal `DbNull` string was stored in the database instead of `NULL`.
+1. You might now encounter a type error or runtime validation error when you use `Prisma.DbNull`, `Prisma.JsonNull`, or `Prisma.AnyNull` with MongoDB. This was never valid, but was silently accepted prior to Prisma 4. You need to review your data and change these fields to `null`.
+1. If you pass in dynamic JSON to a JSON column in Prisma Client (for example `prisma.findMany({where: { jsonColumn: someJson } })`), then you must check that `someJson`cannot be the string "DBNull", "JsonNull", or "AnyNull". If it is any of these values, then the query will return different results in version 4.0.0.
+
+### Default fields on composite types in MongoDB
+
+From version 4.0.0, if you carry out a database read on a composite type when all of the following conditions are true, then Prisma Client inserts the default value into the result.
+
+Conditions:
+
+- A field on the composite type is required, and
+- this field has a default value, and
+- this field is not present in the returned document or documents.
+
+This behavior is now consistent with the behavior for model fields.
+
+To learn more, see [Default values on required composite types](/concepts/components/prisma-client/composite-types#default-values-on-required-composite-types).
```

Comment:
```suggestion
To learn more, see [default values on required composite types](/concepts/components/prisma-client/composite-types#default-values-on-required-composite-types).
```

Are we intentionally capitalizing document titles? If so, feel free to disregard this suggestion and the above suggestion.

Reply:
It's intentional